File: backend/app/services/reports.py
import json
from datetime import date
from threading import Lock

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.constants import ReportStatus, ReportType, TaskStatus
from app.default_templates import DEFAULT_TEMPLATES
from app.models import GenerationTask, LLMSetting, Report, Template, WorkLog, utcnow
from app.services.llm import LLMClient, format_logs_for_prompt
from app.services.periods import format_local_datetime, resolve_period
from app.services.templates import render_template, requires_llm_template_fill, validate_template_content


REPORT_TITLES = {
    ReportType.WEEKLY: "周报",
    ReportType.MONTHLY: "月报",
    ReportType.PERFORMANCE: "绩效考核表",
}
_ACTIVE_GENERATION_KEYS: set[tuple[str, date, date]] = set()
_ACTIVE_GENERATION_LOCK = Lock()


class TemplateList(list):
    """A template value that supports both Markdown output and Jinja iteration."""

    def __init__(self, values: list[object], markdown: str):
        super().__init__(values)
        self.markdown = markdown

    def __str__(self) -> str:
        return self.markdown


def seed_default_templates(db: Session) -> None:
    for template_type, default in DEFAULT_TEMPLATES.items():
        existing = db.scalar(
            select(Template).where(
                Template.template_type == template_type.value,
                Template.is_default.is_(True),
            )
        )
        if existing:
            continue
        db.add(
            Template(
                name=default["name"],
                template_type=template_type.value,
                content=default["content"],
                is_default=True,
            )
        )
    db.commit()


def list_work_logs_for_period(db: Session, start: date, end: date) -> list[WorkLog]:
    return list(
        db.scalars(
            select(WorkLog)
            .where(WorkLog.end_date >= start, WorkLog.start_date <= end)
            .order_by(WorkLog.start_date.asc(), WorkLog.end_date.asc(), WorkLog.id.asc())
        )
    )


def get_template_for_report(db: Session, report_type: ReportType, template_id: int | None) -> Template:
    if template_id:
        template = db.get(Template, template_id)
        if not template or template.template_type != report_type.value:
            raise ValueError("Template not found for report type")
        return template

    template = db.scalar(
        select(Template)
        .where(Template.template_type == report_type.value)
        .order_by(Template.is_default.desc(), Template.updated_at.desc())
    )
    if not template:
        raise ValueError("No template exists for report type")
    return template


def active_llm_setting(db: Session) -> LLMSetting | None:
    return db.scalar(select(LLMSetting).where(LLMSetting.is_active.is_(True)).order_by(LLMSetting.id.desc()))


def report_to_dict_source_ids(report: Report) -> list[int]:
    try:
        return list(json.loads(report.source_log_ids or "[]"))
    except json.JSONDecodeError:
        return []


def build_report_context(
    report_type: ReportType,
    period: tuple[date, date],
    work_logs: list[WorkLog],
    ai_content: str,
) -> dict[str, object]:
    title = f"{period[0]} 至 {period[1]} {REPORT_TITLES[report_type]}"
    results = [item.result for item in work_logs if item.result]
    blockers = [item.blockers for item in work_logs if item.blockers]
    work_item_rows: list[object] = []
    for item in work_logs:
        date_label = (
            str(item.start_date)
            if item.start_date == item.end_date
            else f"{item.start_date} 至 {item.end_date}"
        )
        work_item_rows.append(
            {
                "date": date_label,
                "start_date": str(item.start_date),
                "end_date": str(item.end_date),
                "project": item.project,
                "task": item.task,
                "status": item.progress,
                "content": f"[{item.project}] {item.task}",
                "progress": item.progress,
                "conclusion": item.result or item.notes or "暂无",
                "result": item.result or "",
                "blockers": item.blockers or "",
                "hours": item.hours,
                "priority": item.priority,
                "notes": item.notes or "",
            }
        )
    work_items = TemplateList(
        work_item_rows,
        format_logs_for_prompt(work_logs) or "- 暂无工作记录",
    )
    highlights = TemplateList(results, "\n".join(f"- {item}" for item in results) or "- 暂无")
    blocker_items = TemplateList(blockers, "\n".join(f"- {item}" for item in blockers) or "- 暂无")
    next_step_items = ["继续推进未完成事项"]
    next_steps = TemplateList(next_step_items, "\n".join(f"- {item}" for item in next_step_items))
    return {
        "title": title,
        "report_type": report_type.value,
        "period_start": str(period[0]),
        "period_end": str(period[1]),
        "generated_at": format_local_datetime(utcnow()),
        "ai_content": ai_content,
        "summary": ai_content,
        "work_items": work_items,
        "highlights": highlights,
        "blockers": blocker_items,
        "next_steps": next_steps,
        "raw_llm_content": ai_content,
    }


def create_report(
    db: Session,
    report_type: ReportType,
    anchor_date: date | None = None,
    period_start: date | None = None,
    period_end: date | None = None,
    template_id: int | None = None,
    overwrite: bool = False,
    llm_client: LLMClient | None = None,
) -> tuple[Report, GenerationTask, bool]:
    period = resolve_period(report_type, anchor_date, period_start, period_end)
    existing = db.scalar(
        select(Report).where(
            Report.report_type == report_type.value,
            Report.period_start == period[0],
            Report.period_end == period[1],
        )
    )
    if existing and not overwrite:
        task = GenerationTask(
            report_type=report_type.value,
            period_start=period[0],
            period_end=period[1],
            status=TaskStatus.SKIPPED.value,
            message="Report already exists; existing draft was not overwritten.",
            report_id=existing.id,
            completed_at=utcnow(),
        )
        db.add(task)
        db.commit()
        db.refresh(task)
        return existing, task, False

    generation_key = (report_type.value, period[0], period[1])
    with _ACTIVE_GENERATION_LOCK:
        if generation_key in _ACTIVE_GENERATION_KEYS:
            raise ValueError("Report generation is already running for this period")
        _ACTIVE_GENERATION_KEYS.add(generation_key)

    task: GenerationTask | None = None
    try:
        pending_task = db.scalar(
            select(GenerationTask).where(
                GenerationTask.report_type == report_type.value,
                GenerationTask.period_start == period[0],
                GenerationTask.period_end == period[1],
                GenerationTask.status == TaskStatus.PENDING.value,
            )
        )
        if pending_task:
            raise ValueError("Report generation is already running for this period")

        task = GenerationTask(
            report_type=report_type.value,
            period_start=period[0],
            period_end=period[1],
            status=TaskStatus.PENDING.value,
        )
        db.add(task)
        db.commit()
        db.refresh(task)

        template = get_template_for_report(db, report_type, template_id)
        validate_template_content(template.content)
        logs = list_work_logs_for_period(db, period[0], period[1])
        client = llm_client or LLMClient()
        setting = active_llm_setting(db)
        if requires_llm_template_fill(template.content):
            result = client.fill_template(
                setting,
                REPORT_TITLES[report_type],
                period,
                logs,
                template.content,
            )
            rendered = result.content
            title = f"{period[0]} 至 {period[1]} {REPORT_TITLES[report_type]}"
        else:
            result = client.generate(setting, REPORT_TITLES[report_type], period, logs)
            context = build_report_context(report_type, period, logs, result.content)
            rendered = render_template(template.content, context)
            title = str(context["title"])
        source_log_ids = json.dumps([item.id for item in logs])

        if existing:
            existing.title = title
            existing.template_id = template.id
            existing.content_markdown = rendered
            existing.status = ReportStatus.DRAFT.value
            existing.source_log_ids = source_log_ids
            existing.generated_at = utcnow()
            report = existing
        else:
            report = Report(
                report_type=report_type.value,
                title=title,
                period_start=period[0],
                period_end=period[1],
                template_id=template.id,
                content_markdown=rendered,
                status=ReportStatus.DRAFT.value,
                source_log_ids=source_log_ids,
                generated_at=utcnow(),
            )
            db.add(report)

        db.flush()
        task.status = TaskStatus.SUCCESS.value
        task.report_id = report.id
        task.completed_at = utcnow()
        db.commit()
        db.refresh(report)
        db.refresh(task)
        return report, task, result.used_llm
    except Exception as exc:
        db.rollback()
        failed_task = db.get(GenerationTask, task.id) if task else None
        if failed_task and failed_task.status == TaskStatus.PENDING.value:
            failed_task.status = TaskStatus.FAILED.value
            failed_task.message = str(exc)[:1000]
            failed_task.completed_at = utcnow()
            db.commit()
        raise
    finally:
        with _ACTIVE_GENERATION_LOCK:
            _ACTIVE_GENERATION_KEYS.discard(generation_key)


def create_missing_report(db: Session, report_type: ReportType, anchor_date: date) -> Report | None:
    period = resolve_period(report_type, anchor_date)
    existing = db.scalar(
        select(Report).where(
            Report.report_type == report_type.value,
            Report.period_start == period[0],
            Report.period_end == period[1],
        )
    )
    if existing:
        return None
    report, _, _ = create_report(db, report_type=report_type, anchor_date=anchor_date, overwrite=False)
    return report

File: src-tauri/src/reports.rs
use crate::db::now_string;
use crate::error::{AppError, AppResult};
use crate::llm::{active_config, fill_template, generate_report};
use crate::models::{GenerateResponse, Report, ReportGenerateInput, ReportRow, Template, WorkLog};
use crate::secrets::SecretStore;
use crate::templates::{
    build_context, render_template, report_title, requires_llm_fill, validate_template,
};
use chrono::{Datelike, Duration, Local, NaiveDate, Weekday};
use sqlx::SqlitePool;
use std::collections::HashSet;
use std::sync::Arc;
use tokio::sync::Mutex;

pub type ActiveGenerations = Arc<Mutex<HashSet<String>>>;

pub async fn list_reports(pool: &SqlitePool) -> AppResult<Vec<Report>> {
    Ok(sqlx::query_as::<_, ReportRow>(
        "SELECT id, report_type, title, period_start, period_end, template_id, content_markdown, status, source_log_ids, generated_at, edited_at, created_at, updated_at \
         FROM reports ORDER BY period_end DESC, id DESC",
    )
    .fetch_all(pool)
    .await?
    .into_iter()
    .map(Into::into)
    .collect())
}

pub async fn load_report(pool: &SqlitePool, id: i64) -> AppResult<Report> {
    let row = sqlx::query_as::<_, ReportRow>(
        "SELECT id, report_type, title, period_start, period_end, template_id, content_markdown, status, source_log_ids, generated_at, edited_at, created_at, updated_at \
         FROM reports WHERE id=?",
    )
    .bind(id)
    .fetch_optional(pool)
    .await?
    .ok_or_else(|| AppError::not_found("Report"))?;
    Ok(row.into())
}

pub async fn generate(
    pool: &SqlitePool,
    secrets: &SecretStore,
    active: &ActiveGenerations,
    input: ReportGenerateInput,
) -> AppResult<GenerateResponse> {
    validate_report_type(&input.report_type)?;
    let (period_start, period_end) = resolve_period(
        &input.report_type,
        input.anchor_date.as_deref(),
        input.period_start.as_deref(),
        input.period_end.as_deref(),
    )?;
    let key = format!("{}:{period_start}:{period_end}", input.report_type);
    {
        let mut running = active.lock().await;
        if !running.insert(key.clone()) {
            return Err(AppError::new(
                "generation_in_progress",
                "Report generation is already running for this period",
            ));
        }
    }
    let result = generate_inner(
        pool,
        secrets,
        &input.report_type,
        &period_start,
        &period_end,
        input.template_id,
        input.overwrite,
    )
    .await;
    active.lock().await.remove(&key);
    result
}

async fn generate_inner(
    pool: &SqlitePool,
    secrets: &SecretStore,
    report_type: &str,
    period_start: &str,
    period_end: &str,
    template_id: Option<i64>,
    overwrite: bool,
) -> AppResult<GenerateResponse> {
    let existing = sqlx::query_as::<_, ReportRow>(
        "SELECT id, report_type, title, period_start, period_end, template_id, content_markdown, status, source_log_ids, generated_at, edited_at, created_at, updated_at \
         FROM reports WHERE report_type=? AND period_start=? AND period_end=?",
    )
    .bind(report_type)
    .bind(period_start)
    .bind(period_end)
    .fetch_optional(pool)
    .await?;
    if let Some(existing) = existing.as_ref().filter(|_| !overwrite) {
        let now = now_string();
        let task_id = sqlx::query(
            "INSERT INTO generation_tasks(report_type, period_start, period_end, status, message, report_id, completed_at, created_at, updated_at) \
             VALUES(?, ?, ?, 'skipped', 'Report already exists; existing draft was not overwritten.', ?, ?, ?, ?)",
        )
        .bind(report_type)
        .bind(period_start)
        .bind(period_end)
        .bind(existing.id)
        .bind(&now)
        .bind(&now)
        .bind(&now)
        .execute(pool)
        .await?
        .last_insert_rowid();
        return Ok(GenerateResponse {
            report: existing.clone().into(),
            task_id,
            used_llm: false,
        });
    }

    let now = now_string();
    let task_id = sqlx::query(
        "INSERT INTO generation_tasks(report_type, period_start, period_end, status, created_at, updated_at) \
         VALUES(?, ?, ?, 'pending', ?, ?)",
    )
    .bind(report_type)
    .bind(period_start)
    .bind(period_end)
    .bind(&now)
    .bind(&now)
    .execute(pool)
    .await?
    .last_insert_rowid();

    let result: AppResult<GenerateResponse> = async {
        let template = get_template(pool, report_type, template_id).await?;
        validate_template(&template.content)?;
        let logs = sqlx::query_as::<_, WorkLog>(
            "SELECT id, work_date, start_date, end_date, project, task, progress, result, blockers, hours, priority, notes, created_at, updated_at \
             FROM work_logs WHERE end_date >= ? AND start_date <= ? ORDER BY start_date, end_date, id",
        )
        .bind(period_start)
        .bind(period_end)
        .fetch_all(pool)
        .await?;
        let config = active_config(pool, secrets).await?;
        let (content, used_llm, title) = if requires_llm_fill(&template.content) {
            let generated = fill_template(
                config.as_ref(),
                report_type,
                period_start,
                period_end,
                &logs,
                &template.content,
            )
            .await?;
            (
                generated.content,
                generated.used_llm,
                format!("{} 至 {} {}", period_start, period_end, report_title(report_type)),
            )
        } else {
            let generated = generate_report(
                config.as_ref(),
                report_type,
                period_start,
                period_end,
                &logs,
            )
            .await?;
            let context = build_context(
                report_type,
                period_start,
                period_end,
                &logs,
                generated.content,
                local_datetime(),
            );
            let content = render_template(&template.content, &context)?;
            (content, generated.used_llm, context.title)
        };
        let source_ids = serde_json::to_string(&logs.iter().map(|item| item.id).collect::<Vec<_>>())?;
        let timestamp = now_string();
        let report_id = if let Some(existing) = existing {
            sqlx::query(
                "UPDATE reports SET title=?, template_id=?, content_markdown=?, status='draft', source_log_ids=?, generated_at=?, updated_at=? WHERE id=?",
            )
            .bind(&title)
            .bind(template.id)
            .bind(&content)
            .bind(&source_ids)
            .bind(&timestamp)
            .bind(&timestamp)
            .bind(existing.id)
            .execute(pool)
            .await?;
            existing.id
        } else {
            sqlx::query(
                "INSERT INTO reports(report_type, title, period_start, period_end, template_id, content_markdown, status, source_log_ids, generated_at, created_at, updated_at) \
                 VALUES(?, ?, ?, ?, ?, ?, 'draft', ?, ?, ?, ?)",
            )
            .bind(report_type)
            .bind(&title)
            .bind(period_start)
            .bind(period_end)
            .bind(template.id)
            .bind(&content)
            .bind(&source_ids)
            .bind(&timestamp)
            .bind(&timestamp)
            .bind(&timestamp)
            .execute(pool)
            .await?
            .last_insert_rowid()
        };
        sqlx::query(
            "UPDATE generation_tasks SET status='success', report_id=?, completed_at=?, updated_at=? WHERE id=?",
        )
        .bind(report_id)
        .bind(&timestamp)
        .bind(&timestamp)
        .bind(task_id)
        .execute(pool)
        .await?;
        Ok(GenerateResponse {
            report: load_report(pool, report_id).await?,
            task_id,
            used_llm,
        })
    }
    .await;

    if let Err(error) = &result {
        let now = now_string();
        let _ = sqlx::query(
            "UPDATE generation_tasks SET status='failed', message=?, completed_at=?, updated_at=? WHERE id=?",
        )
        .bind(error.message.chars().take(1000).collect::<String>())
        .bind(&now)
        .bind(&now)
        .bind(task_id)
        .execute(pool)
        .await;
    }
    result
}

async fn get_template(
    pool: &SqlitePool,
    report_type: &str,
    template_id: Option<i64>,
) -> AppResult<Template> {
    let template = if let Some(id) = template_id {
        sqlx::query_as::<_, Template>(
            "SELECT id, name, template_type, content, is_default, created_at, updated_at FROM templates WHERE id=? AND template_type=?",
        )
        .bind(id)
        .bind(report_type)
        .fetch_optional(pool)
        .await?
    } else {
        sqlx::query_as::<_, Template>(
            "SELECT id, name, template_type, content, is_default, created_at, updated_at FROM templates WHERE template_type=? ORDER BY is_default DESC, updated_at DESC LIMIT 1",
        )
        .bind(report_type)
        .fetch_optional(pool)
        .await?
    };
    template
        .ok_or_else(|| AppError::new("template_not_found", "No template exists for report type"))
}

pub fn resolve_period(
    report_type: &str,
    anchor: Option<&str>,
    start: Option<&str>,
    end: Option<&str>,
) -> AppResult<(String, String)> {
    validate_report_type(report_type)?;
    if let (Some(start), Some(end)) = (start, end) {
        let start_date = parse_date(start, "period_start")?;
        let end_date = parse_date(end, "period_end")?;
        if end_date < start_date {
            return Err(AppError::validation(
                "period_end",
                "period_end must be on or after period_start",
            ));
        }
        return Ok((start.into(), end.into()));
    }
    let anchor = anchor
        .map(|value| parse_date(value, "anchor_date"))
        .transpose()?
        .unwrap_or_else(|| Local::now().date_naive());
    let (start, end) = if report_type == "weekly_report" {
        let offset = anchor.weekday().num_days_from_monday() as i64;
        let start = anchor - Duration::days(offset);
        (start, start + Duration::days(6))
    } else {
        let start = anchor.with_day(1).unwrap();
        let next_month = if anchor.month() == 12 {
            NaiveDate::from_ymd_opt(anchor.year() + 1, 1, 1).unwrap()
        } else {
            NaiveDate::from_ymd_opt(anchor.year(), anchor.month() + 1, 1).unwrap()
        };
        (start, next_month - Duration::days(1))
    };
    Ok((start.to_string(), end.to_string()))
}

pub fn scheduled_period(report_type: &str, occurrence: NaiveDate) -> AppResult<(String, String)> {
    validate_report_type(report_type)?;
    if report_type == "weekly_report" {
        let start = occurrence - Duration::days(occurrence.weekday().num_days_from_monday() as i64);
        Ok((start.to_string(), occurrence.to_string()))
    } else {
        Ok((
            occurrence.with_day(1).unwrap().to_string(),
            occurrence.to_string(),
        ))
    }
}

pub fn validate_report_type(value: &str) -> AppResult<()> {
    match value {
        "weekly_report" | "monthly_report" | "performance_review" => Ok(()),
        _ => Err(AppError::validation(
            "report_type",
            "Unsupported report type",
        )),
    }
}

pub fn parse_date(value: &str, field: &str) -> AppResult<NaiveDate> {
    NaiveDate::parse_from_str(value, "%Y-%m-%d")
        .map_err(|_| AppError::validation(field, "Date must use YYYY-MM-DD format"))
}

fn local_datetime() -> String {
    chrono::Utc::now()
        .with_timezone(&chrono_tz::Asia::Shanghai)
        .format("%Y-%m-%d %H:%M:%S")
        .to_string()
}

pub fn weekday_from_code(value: &str) -> Option<Weekday> {
    match value {
        "mon" => Some(Weekday::Mon),
        "tue" => Some(Weekday::Tue),
        "wed" => Some(Weekday::Wed),
        "thu" => Some(Weekday::Thu),
        "fri" => Some(Weekday::Fri),
        "sat" => Some(Weekday::Sat),
        "sun" => Some(Weekday::Sun),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::db::{now_string, Database};
    use crate::secrets::SecretStore;
    use std::collections::HashSet;
    use std::sync::Arc;
    use tokio::sync::Mutex;

    #[test]
    fn resolves_week_and_month_periods() {
        assert_eq!(
            resolve_period("weekly_report", Some("2026-06-26"), None, None).unwrap(),
            ("2026-06-22".into(), "2026-06-28".into())
        );
        assert_eq!(
            resolve_period("monthly_report", Some("2026-06-12"), None, None).unwrap(),
            ("2026-06-01".into(), "2026-06-30".into())
        );
    }

    #[test]
    fn scheduled_week_ends_on_occurrence() {
        assert_eq!(
            scheduled_period(
                "weekly_report",
                NaiveDate::from_ymd_opt(2026, 6, 26).unwrap()
            )
            .unwrap(),
            ("2026-06-22".into(), "2026-06-26".into())
        );
    }

    #[tokio::test]
    async fn generates_a_fallback_report_without_an_llm_key() {
        let unique = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_nanos();
        let directory = std::env::temp_dir().join(format!(
            "worklog-report-test-{}-{unique}",
            std::process::id()
        ));
        std::fs::create_dir_all(&directory).unwrap();
        let database = Database::open_for_test(directory, None).await.unwrap();
        let now = now_string();
        sqlx::query(
            "INSERT INTO work_logs(work_date, start_date, end_date, project, task, progress, result, hours, priority, created_at, updated_at) \
             VALUES('2026-06-23', '2026-06-23', '2026-06-23', 'Worklog', 'Rust desktop', '完成迁移', '可本地运行', 4, 'high', ?, ?)",
        )
        .bind(&now)
        .bind(&now)
        .execute(&database.pool)
        .await
        .unwrap();
        let generated = generate(
            &database.pool,
            &SecretStore::memory(),
            &Arc::new(Mutex::new(HashSet::new())),
            ReportGenerateInput {
                report_type: "monthly_report".into(),
                anchor_date: Some("2026-06-23".into()),
                period_start: None,
                period_end: None,
                template_id: None,
                overwrite: false,
            },
        )
        .await
        .unwrap();
        assert!(!generated.used_llm);
        assert!(generated.report.content_markdown.contains("Rust desktop"));
        assert_eq!(generated.report.period_start, "2026-06-01");
        assert_eq!(generated.report.period_end, "2026-06-30");

        sqlx::query("UPDATE reports SET content_markdown='# 手工修改后不可覆盖' WHERE id=?")
            .bind(generated.report.id)
            .execute(&database.pool)
            .await
            .unwrap();
        let duplicate = generate(
            &database.pool,
            &SecretStore::memory(),
            &Arc::new(Mutex::new(HashSet::new())),
            ReportGenerateInput {
                report_type: "monthly_report".into(),
                anchor_date: Some("2026-06-23".into()),
                period_start: None,
                period_end: None,
                template_id: None,
                overwrite: false,
            },
        )
        .await
        .unwrap();
        let task_status: String =
            sqlx::query_scalar("SELECT status FROM generation_tasks WHERE id=?")
                .bind(duplicate.task_id)
                .fetch_one(&database.pool)
                .await
                .unwrap();
        assert_eq!(task_status, "skipped");
        assert_eq!(duplicate.report.content_markdown, "# 手工修改后不可覆盖");
    }
}
